torchtune/modules/kv_cache.py

```diff
@@ -64,11 +64,10 @@ def update(
             Tuple[Tensor, Tensor]: Updated KV cache with key first
         """
         assert input_pos.shape[0] == k_val.shape[2]
-        self.size = input_pos.max().item() + 1
 
         k_out = self.k_cache
         v_out = self.v_cache
-        k_out[:, :, input_pos] = k_val
-        v_out[:, :, input_pos] = v_val
```

Comment:
question: We need to do it because the base model has one dtype, e.g. float32, but the cache is in fp16, for example, is that right?

after we save the k_val and v_val in fp16, we have to return the entire cache to be used by the attention. Dont we have to upcast it back to float32, since the model is in float32? I dont see it happening anywhere.

Reply:
The dtype in the yaml config is actually bfloat16 so the model is cast to that dtype [here](https://github.com/pytorch/torchtune/blob/2f8ed7a34457eaddce59a58e8c37dd51d6ba65f8/recipes/generate.py#L73). Its actually the quantization component wrapping the linear layers that returns float32, so we need to explicitly downcast to the same dtype as the model. See [here ](https://github.com/pytorch/torchtune/issues/1349#issuecomment-2297562155) why this returns float32. I think we would need to upstream changes to torchao to change this.

Alternatively, we could do the downcast right after computing q,k,v but kv cache already had the dtype given during initialization. Otherwise we either need derive the dtype for q,k,v from the model parameters or change the interface for the attention layer creation to accept a dtype.

Reply:
>Its actually the quantization component wrapping the linear layers that returns float32

@mreso, thanks for explaining it! Lets try to fix the root cause then. I dont think it is the intended behavior of the quantizer to return float32 if the input is float16.[ I took a look at the wrapper ](https://github.com/pytorch/ao/blob/227d4bfece6bb00f0365642ad1d7afd509079ba4/torchao/quantization/GPTQ.py#L893)and it says that:

"precision: precision of input and output. e.g. torch.float32 means input
activation is float32 and output is float32.
scales_precision: precision of per group scale."

If you have bandwidth, would you mind doing a quick test with the quantizer dtype inputs? If the inputs is indeed fp16, and precision=fp16 (and scales_precision), and it still output fp32, then I will ping the torchao folks.

Reply:
Here is a quick test:
```
import torch
import torch.nn as nn
from torchtune.utils.quantization import Int8DynActInt4WeightQuantizer


class Foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(256, 256, bias=False)
        self.linear2 = nn.Linear(256, 256, bias=False)

    def forward(self, x):
        x = self.linear(x)
        x = self.linear2(x)
        return x



quantizer_f32 = Int8DynActInt4WeightQuantizer()

foo = Foo()
foo = quantizer_f32.quantize(foo)
foo = foo.to(dtype=torch.bfloat16)

# foo = torch.compile(foo)

x_bf16 = torch.rand(1,256, dtype=torch.bfloat16)
print(f"{x_bf16.dtype=}")
y=foo(x_bf16)
print(f"{y.dtype=}")

x_f32 = torch.rand(1,256, dtype=torch.float32)
print(f"{x_f32.dtype=}")
y=foo(x_f32)
print(f"{y.dtype=}")

quantizer_bf16 = Int8DynActInt4WeightQuantizer(precision=torch.bfloat16)

bar = Foo()
bar = quantizer_bf16.quantize(bar)
bar = bar.to(dtype=torch.bfloat16)

# bar = torch.compile(bar)

x = torch.rand(1,256, dtype=torch.bfloat16)
print(f"{x.dtype=}")
y=bar(x)
print(f"{y.dtype=}")
```
As the docs state the output's precision depends on quantizer.precision. So for bfloat16 in it should be bfloat16 out:
```
linear: linear, in=256, out=256
linear: linear2, in=256, out=256
x_bf16.dtype=torch.bfloat16
y.dtype=torch.float32
x_f32.dtype=torch.float32
y.dtype=torch.float32
linear: linear, in=256, out=256
linear: linear2, in=256, out=256
x.dtype=torch.bfloat16
y.dtype=torch.bfloat16
```
This brings us back to two problems described in the original issue:
1. We currently can not set the precision with the current implementation of ao + torchtune as precision has to be a dtype but torchtune only passes the string read from the config.
2. If we somehow set precision to the correct dtype as e.g. [described here](https://github.com/pytorch/torchtune/issues/1349#issuecomment-2294317411) AND enable torch.compile (which the generate recipe does) we end up with the inductor error.
```
Traceback (most recent call last):
  File "/home/mreso/torchtune/test_quantizer.py", line 47, in <module>
    y=bar(x)
 ...
  File "/tmp/torchinductor_mreso/ji/cjib4exg2swki3pwhhi7cgq5qh7l6yriypsqejd3wdcajrx55zrc.py", line 69, in call
    extern_kernels.mm(buf6, reinterpret_tensor(buf8, (256, 256), (1, 256), 0), out=buf9)
RuntimeError: Expected out tensor to have dtype float, but got c10::BFloat16 instead
```
So if we don't want to solve the problem by casting we need to fix the inductor issue on the ao side.

Reply:
Nice!

1. one is easy to fix. We can just use [this mapping](https://github.com/pytorch/torchtune/blob/34162c9b2b24e6e555f6a06074dc0c34431a5aa8/torchtune/utils/precision.py#L17) to go from string to torch.bfloat16


2. This seems to need attention

@msaroufim, can you take a look at this issue? 

**TLDR:**
using compile + Int8DynActInt4WeightQuantizer with precision=bf16 breaks.
```
quantizer_bf16 = Int8DynActInt4WeightQuantizer(precision=torch.bfloat16)

bar = Foo()
bar = quantizer_bf16.quantize(bar)
bar = bar.to(dtype=torch.bfloat16)
bar = torch.compile(bar)

x = torch.rand(1,256, dtype=torch.bfloat16)
y=bar(x)
```

raises the error:
```
raceback (most recent call last):
  File "/home/mreso/torchtune/test_quantizer.py", line 47, in <module>
    y=bar(x)
 ...
  File "/tmp/torchinductor_mreso/ji/cjib4exg2swki3pwhhi7cgq5qh7l6yriypsqejd3wdcajrx55zrc.py", line 69, in call
    extern_kernels.mm(buf6, reinterpret_tensor(buf8, (256, 256), (1, 256), 0), out=buf9)
RuntimeError: Expected out tensor to have dtype float, but got c10::BFloat16 instead
```

If we dont add ```precision=torch.bfloat16```, then for an fp16 input, we get fp32 output.

Reply:
@jerryzh168 any reason why we're not using `quantize_()` here instead?

Reply:
For reference, here is where the component is defined in the config: https://github.com/pytorch/torchtune/blob/main/recipes/configs/quantization.yaml#L27

Reply:
And here is where its instantiated: https://github.com/pytorch/torchtune/blob/9e65fa9ae67ce09e953cb68b5093d42cb6b31310/recipes/quantize.py#L51

Reply:
@msaroufim there was some issues with loading the tensor subclass quantized weights in `generate.py` before (with map_location="cpu": https://github.com/pytorch/torchtune/blob/f9f75bb563ecae371492a9d49da4a9f514c081b3/torchtune/utils/_checkpointing/_checkpointer_utils.py#L110-L111), things might have changed though. Also we need to clarify the plan for QAT flow I think. cc @andrewor14 is the recent tensor subclass refactor tested in torchtune?

Reply:
@msaroufim @jerryzh168 @felipemello1 is there any short term solution to unblock us here? If its the original issue or the PR?